File: src/asr/AsrWhisper.ts
import { logger } from '../utils/logging.js';

export interface AsrConfig {
  modelSize: 'tiny' | 'base' | 'small' | 'medium' | 'large';
  language?: string;
  temperature?: number;
  initialPrompt?: string;
}

export interface AsrResult {
  text: string;
  confidence: number;
  language: string;
  duration: number;
  timestamp: number;
}

export class AsrWhisper {
  private config: AsrConfig;
  private whisper: any = null; // Will be loaded dynamically

  constructor(config: AsrConfig) {
    this.config = {
      modelSize: 'base',
      language: 'en',
      temperature: 0.0,
      initialPrompt: 'Hello, I am JARVIS. How may I help you?',
      ...config
    };
  }

  async initialize(): Promise<void> {
    try {
      // Dynamic import of whisper (assuming whisper-node or similar)
      // Note: This would need actual whisper integration
      logger.info('Initializing Whisper ASR...', {
        modelSize: this.config.modelSize,
        language: this.config.language
      });
      
      // Placeholder for actual whisper initialization
      // In production, this would load the actual Whisper model
      this.whisper = {
        transcribe: this.mockTranscribe.bind(this)
      };
      
      logger.info('Whisper ASR initialized successfully');
      
    } catch (error) {
      logger.error('Failed to initialize Whisper:', error);
      throw new Error(`Whisper initialization failed: ${error}`);
    }
  }

  async transcribe(audioData: Float32Array, sampleRate: number = 16000): Promise<AsrResult> {
    if (!this.whisper) {
      throw new Error('Whisper not initialized');
    }

    const startTime = performance.now();
    
    try {
      logger.debug(`Transcribing audio: ${audioData.length} samples at ${sampleRate}Hz`);
      
      // Validate audio data
      if (audioData.length === 0) {
        throw new Error('Empty audio data');
      }
      
      // Check for minimum audio length (e.g., 0.1 seconds)
      const minSamples = sampleRate * 0.1;
      if (audioData.length < minSamples) {
        logger.warn('Audio too short for reliable transcription');
      }
      
      // Calculate audio level to detect silence
      const audioLevel = this.calculateAudioLevel(audioData);
      if (audioLevel < 0.001) {
        logger.debug('Audio appears to be mostly silence');
        return {
          text: '',
          confidence: 0.0,
          language: this.config.language || 'en',
          duration: performance.now() - startTime,
          timestamp: Date.now()
        };
      }
      
      // Perform transcription
      const result = await this.whisper.transcribe(audioData, {
        language: this.config.language,
        temperature: this.config.temperature,
        initialPrompt: this.config.initialPrompt,
        sampleRate
      });
      
      const duration = performance.now() - startTime;
      
      logger.debug(`Transcription completed in ${duration.toFixed(1)}ms: "${result.text}"`);
      
      return {
        text: result.text.trim(),
        confidence: result.confidence || 0.8,
        language: result.language || this.config.language || 'en',
        duration,
        timestamp: Date.now()
      };
      
    } catch (error) {
      logger.error('Transcription failed:', error);
      throw error;
    }
  }

  private calculateAudioLevel(audioData: Float32Array): number {
    let sum = 0;
    for (let i = 0; i < audioData.length; i++) {
      sum += Math.abs(audioData[i]);
    }
    return sum / audioData.length;
  }

  // Mock transcribe for development/testing
  private async mockTranscribe(audioData: Float32Array, options: any): Promise<any> {
    // Simulate processing time
    await new Promise(resolve => setTimeout(resolve, 100 + Math.random() * 400));
    
    const audioLevel = this.calculateAudioLevel(audioData);
    
    if (audioLevel < 0.001) {
      return { text: '', confidence: 0.0, language: 'en' };
    }
    
    // Mock responses based on audio characteristics
    const responses = [
      'Hello JARVIS',
      'What time is it',
      'Turn on the lights',
      'Play some music',
      'What\'s the weather like',
      'Set a timer for 5 minutes',
      'How are you doing',
      'Thank you'
    ];
    
    const randomResponse = responses[Math.floor(Math.random() * responses.length)];
    
    return {
      text: randomResponse,
      confidence: 0.8 + Math.random() * 0.2,
      language: 'en'
    };
  }

  // Convert audio format if needed
  convertToFloat32(audioData: Int16Array): Float32Array {
    const float32Data = new Float32Array(audioData.length);
    for (let i = 0; i < audioData.length; i++) {
      float32Data[i] = audioData[i] / 32768.0; // Convert from int16 to float32
    }
    return float32Data;
  }

  // Preprocess audio (e.g., normalize, filter)
  preprocessAudio(audioData: Float32Array): Float32Array {
    // Apply basic normalization
    const maxValue = Math.max(...audioData.map(Math.abs));
    if (maxValue > 0) {
      const scaleFactor = 0.95 / maxValue;
      return audioData.map(sample => sample * scaleFactor);
    }
    return audioData;
  }

  cleanup(): void {
    if (this.whisper) {
      this.whisper = null;
      logger.info('Whisper ASR cleaned up');
    }
  }

  // Get supported languages
  static getSupportedLanguages(): string[] {
    return [
      'en', 'es', 'fr', 'de', 'it', 'pt', 'ru', 'ja', 'ko', 'zh',
      'ar', 'hi', 'tr', 'pl', 'nl', 'sv', 'da', 'no', 'fi'
    ];
  }

  // Get model size recommendations
  static getModelRecommendations(): Record<string, string> {
    return {
      'tiny': 'Fastest, least accurate (~40MB)',
      'base': 'Good balance of speed and accuracy (~150MB)',
      'small': 'Better accuracy, slower (~500MB)',
      'medium': 'High accuracy, much slower (~1.5GB)',
      'large': 'Best accuracy, very slow (~3GB)'
    };
  }
}

File: jarvis/asr/asr_whisper.py
"""Ported from src/asr/AsrWhisper.ts.

Unlike the TypeScript original (which only mocked transcription), this
wraps the real `openai-whisper` package -- the same one already proven
working in wake_system_integrated.py.
"""
from __future__ import annotations

import time
from dataclasses import dataclass
from typing import Any, Optional

import numpy as np

from ..utils.logging import logger

_SUPPORTED_LANGUAGES = [
    "en", "es", "fr", "de", "it", "pt", "ru", "ja", "ko", "zh",
    "ar", "hi", "tr", "pl", "nl", "sv", "da", "no", "fi",
]

_MODEL_RECOMMENDATIONS = {
    "tiny": "Fastest, least accurate (~40MB)",
    "base": "Good balance of speed and accuracy (~150MB)",
    "small": "Better accuracy, slower (~500MB)",
    "medium": "High accuracy, much slower (~1.5GB)",
    "large": "Best accuracy, very slow (~3GB)",
}


@dataclass
class AsrConfig:
    model_size: str = "base"
    language: str = "en"
    temperature: float = 0.0
    initial_prompt: str = "Hello, I am JARVIS. How may I help you?"


@dataclass
class AsrResult:
    text: str
    confidence: float
    language: str
    duration: float
    timestamp: float


class AsrWhisper:
    def __init__(self, config: Optional[AsrConfig] = None) -> None:
        self.config = config or AsrConfig()
        self._model: Any = None

    def initialize(self) -> None:
        import whisper

        logger.info("Initializing Whisper ASR...", {"modelSize": self.config.model_size, "language": self.config.language})
        self._model = whisper.load_model(self.config.model_size)
        logger.info("Whisper ASR initialized successfully")

    def transcribe(self, audio_data: np.ndarray, sample_rate: int = 16000) -> AsrResult:
        if self._model is None:
            raise RuntimeError("Whisper not initialized")

        start = time.perf_counter()
        logger.debug(f"Transcribing audio: {len(audio_data)} samples at {sample_rate}Hz")

        if len(audio_data) == 0:
            raise ValueError("Empty audio data")

        audio_level = self._audio_level(audio_data)
        if audio_level < 0.001:
            logger.debug("Audio appears to be mostly silence")
            return AsrResult("", 0.0, self.config.language, (time.perf_counter() - start) * 1000, time.time() * 1000)

        import whisper

        audio_f32 = audio_data.astype(np.float32)
        peak = np.max(np.abs(audio_f32))
        if peak > 0:
            audio_f32 = audio_f32 / peak
        audio_f32 = whisper.pad_or_trim(audio_f32)

        result = self._model.transcribe(
            audio_f32,
            language=self.config.language,
            fp16=False,
            no_speech_threshold=0.3,
            condition_on_previous_text=False,
            initial_prompt=self.config.initial_prompt,
        )

        duration = (time.perf_counter() - start) * 1000
        text = result["text"].strip()
        logger.debug(f'Transcription completed in {duration:.1f}ms: "{text}"')

        return AsrResult(text, 0.8, result.get("language", self.config.language), duration, time.time() * 1000)

    def _audio_level(self, audio_data: np.ndarray) -> float:
        return float(np.mean(np.abs(audio_data)))

    def convert_to_float32(self, audio_data: np.ndarray) -> np.ndarray:
        return audio_data.astype(np.float32) / 32768.0

    def preprocess_audio(self, audio_data: np.ndarray) -> np.ndarray:
        max_value = float(np.max(np.abs(audio_data))) if len(audio_data) else 0.0
        if max_value > 0:
            return audio_data * (0.95 / max_value)
        return audio_data

    def cleanup(self) -> None:
        self._model = None
        logger.info("Whisper ASR cleaned up")

    @staticmethod
    def get_supported_languages() -> list[str]:
        return list(_SUPPORTED_LANGUAGES)

    @staticmethod
    def get_model_recommendations() -> dict[str, str]:
        return dict(_MODEL_RECOMMENDATIONS)
